[14.0] Add module mail_notification_with_history

TDu opened the pull request:
Followers of a discussion in the chatter of Odoo can be informed by email
of a new message being created.
This module adds in that email the history of the chatter discussion to help
the recipient with more context.

TDu committed:
Add module mail_notification_with_history

Followers of a discussion in the chatter of Odoo can be informed by email
of a new message being created.
This module adds in that email the history of the chatter discussion to help
the recipient with more context.

ivantodorovich commented on mail_notification_with_history/models/mail_thread.py:
```diff
@@ -0,0 +1,14 @@
+    _inherit = "mail.thread"
+
+    _mail_notification_include_history = False
+
+    @api.model
+    def with_mail_notification_history(self):
+        return self._mail_notification_include_history
```
```suggestion
```

I'd remove this and stick to using the property.
Properties can still be overriden, anyway.

btw, public model methods are exposed through xmlrpc.
Rule of thumb is keep everything private unless you really want this, IMO.

TDu replied:
I added that one when I was toying with the mocking in the tests. It is removed

ivantodorovich commented on mail_notification_with_history/models/mail_message.py:
```diff
@@ -0,0 +1,29 @@
+
+    def get_notification_message_history(self):
+        """Get the list of messages to include into an email notification history."""
+        if not self.env[self.model].with_mail_notification_history():
+            return []
+        if not self._with_notification_message_history():
+            return []
```
```suggestion
        if (
            not self.env[self.model].with_mail_notification_history()
            or not self._with_notification_message_history()
        ):
            return self.browse()
```

note that, since you'd normally return a `mail.message` recordset, here we should return an empty recordset for consistency rather than an empty list.

ivantodorovich replied:
Although, as per my comment below, I'd go for:

```suggestion
        if (
            not self.env[self.model]._mail_notification_include_history
            or not self._with_notification_message_history()
        ):
            return self.browse()
```

TDu replied:
Yes for the browse instead of the list.

ivantodorovich commented on mail_notification_with_history/models/mail_message.py:
```diff
@@ -0,0 +1,29 @@
+            ("message_type", "in", ("user_notification", "comment", "email")),
+        ]
+
+    def _with_notification_message_history(self):
```
```suggestion
    def _should_include_notification_message_history(self):
```

Just a suggestion. Feel free to ignore it.

IMO `with_*` implies some kind of change in the recordset. Makes me think of `with_company`, `with_user`, `with_prefetch`...

Also probably as a hook point doesn't provide much value because one could simply override `_get_notification_message_history` to achieve the same 😓 so could even remove it

TDu replied:
Well I removed this one as it is not really needed.

ivantodorovich commented on mail_notification_with_history/models/mail_message.py:
```diff
@@ -0,0 +1,29 @@
+class Message(models.Model):
+    _inherit = "mail.message"
+
+    def get_notification_message_history(self):
```
```suggestion
    def _get_notification_message_history(self):
```
See comment below about public methods

TDu replied:
changed.

ivantodorovich commented on mail_notification_with_history/models/mail_message.py:
```diff
@@ -0,0 +1,29 @@
+        if not self._with_notification_message_history():
+            return []
+        domain = self._get_notification_message_history_domain()
+        messages = self.env["mail.message"].search(domain, order="date desc")
```
How about also implementing a limit? It could be important in case there are long discussions
Could be overrideable with a method `_get_notification_message_history_limit` in the `mail.thread` (see comment below about `_get_notification_message_history_domain`)

TDu replied:
About the limit of messages lets see with the use of the module if this becomes a requirement.

ivantodorovich commented on mail_notification_with_history/models/mail_message.py:
```diff
@@ -0,0 +1,29 @@
+        messages = self.env["mail.message"].search(domain, order="date desc")
+        return messages - self
+
+    def _get_notification_message_history_domain(self):
```
I'd move this method to `mail.thread`, so that it can be overriden per-model

TDu replied:
Not sure...

ivantodorovich reviewed (approved):
Thanks ! LG !

TDu commented:
@ivantodorovich Thank you for your reviews.
I have slightly change the domain that search for message to include. As to include only the comment that are not internal note. And added a 2nd test.
Also squashed the fixups.

dreispt commented:
/ocabot merge nobump
